### DISCUSSION TITLE
Counter rule

### DISCUSSION THREAD

**Discussion by DutchCoban** [arendst/Tasmota#14257] [Q&A] [answered]
I've connected a LJ18A3-8-Z/BX-5V (proximity sensor) to a Sonoff Basic R2 so I can read my watermeter (Sensus 620) and it works like a charm. I've set the counter to a certain value and since then Counter value C1 is the same as the value on my watermeter. After I had to restore a backup of Home Assistant I lost the waterusage of that day.

Is there a rule I can use so Tasmota publishes both the totalizer value (C1) and a daily value.
The general idea is:
Rule1 If C1 is increased by x then VAR1 (or whatever) should increase the same amount.
Rule2 At 0:00 VAR1 should be zero'd.

Any help would be appreciated.

**Comment by sfromis** [comment, marked as answer]
Tasmota has no feature for saving a counter value, but you could do something with rules. What I do in my backend is to make sure to not lose data, meaning that the max loss would be since last TelePeriod.

**Reply by DutchCoban** [reply]
I have a different oppinion about the counter value saving, but there might be a miscommunication here. I've set the initial value by entering "COUNTER1 [initial value]" in the console of my Sonoff Basic, from that point every pulse the proximity sensor generates is being added to this value. The countervalue C1 has survided many restarts and powerlosses so far. If the problem is that variables wouldn't survive a restart, then I can assisgn a counter (C2) to GPIO2.

**Reply by sfromis** [reply]
The counter can survive scheduled restarts of Tasmota, and it will be saved when the settings area is saved for other reasons. I've not tested if it is saved at midnight too, like the energy monitoring totals, but that would be consistent. What cannot be saved is pulses since last time the settings data were saved, in case of a "restart" meaning power cycling. Saving is very deliberately not done every time, as the flash chip endurance only allows a limited number of writes.

If the settings save did not work for whatever reason (including power or hardware issues), Tasmota will also return to "last known good", excluding updates since then, instead of just going dead.

Whatever happened when you restored that HA backup is well outside of Tasmota, and may also depend on whatever automations you may have, purely speculatively like if you have an automation to resend last known counter data from HA to Tasmota when deemed relevant.

**Comment by DutchCoban** [comment]
As stuborn as I am I've assigned a counter (2) to GPIO2.
I'd prefer to hardwire GPIO3 (Counter1, C1) to GPIO2, but that will probably lead to some errors in case the device is booting.
Alternatively I've created the following rule:
**RULE1 ON Counter#C1 DO Counter2 +1 ENDON**
As for my understanding of the rules, whenever Counter#C1 is increased, Counter2 should increase as well.
Unfortunately my thoughts where to simplistic perhaps because Counter C2 is continuously increasing, even if C1 isn't.

**Counter2 +1** works through the command line, so that part of the rule shouldn't be a problem.
```
14:09:18.418 CMD: counter2 +1
14:09:18.426 MQT: stat/78/RESULT = {"Counter2":1}
14:09:19.572 CMD: counter2 +1
14:09:19.580 MQT: stat/78/RESULT = {"Counter2":2}
14:09:20.503 CMD: counter2 +1
14:09:20.511 MQT: stat/78/RESULT = {"Counter2":3}
14:09:21.348 CMD: counter2 +1
14:09:21.356 MQT: stat/78/RESULT = {"Counter2":4}
14:09:22.151 CMD: counter2 +1
14:09:22.159 MQT: stat/78/RESULT = {"Counter2":5}
14:09:23.037 CMD: counter2 +1
14:09:23.045 MQT: stat/78/RESULT = {"Counter2":6}
```
[Connected sensors can be a trigger in the form as they are represented in the TelePeriod and Status 8 JSON payloads.
Example: DS18B20#Temperature; whenever the temperature of sensor DS18B20 changes](https://tasmota.github.io/docs/Rules/#rule-trigger)

My status 8 reads: 
`STATUS8 = {"StatusSNS":{"Time":"2022-01-04T14:05:09","COUNTER":{"C1":2053087,"C2":0}}}`

**Reply by sfromis** [reply]
You say that "Counter C2 is continuously increasing, even if C1 isn't" and then post `"COUNTER":{"C1":2053087,"C2":0}` 
I see no increase in C2 being zero. However, the rule you post means that you want to increase counter 2 for every cycle (about 1 second) the rule trigger engine is sampling C1, *changed or not*. If you want it to react to changes only, you'll have to make the rule save the value, and the trigger only fire if the new value is different from the old.

Not sure what "that will probably lead to some errors" is really supposed to mean. Or what benefit you see in incrementing two counters at the same time. When adding complexity, better have a clear reasoning for what the gain is.

**Reply by DutchCoban** [reply]
C2 was increasing indeed after every second or so, not with the increase of value C1. I’m going to give the “save value and trigger when changed” a try. I’m learning as we go, so forgive me.

Also on the GPIO1 and 2 part I’m learning, GPIO2 shouldn’t be grounded at boot because it will probably go into bootloader-mode. My LJ18… is putting out 5V so I’m probably not going to have issue here on second thought. GPIO1 also shouldn’t pulled low because it won’t start then. Also this shouldn’t be a problem because the LJ18… is putting out 5V.

Reason for me to have a secondary or daily counter is for the same reason people reading out their electricity usage have one. I’m not reading out electricity usage but total water usage and I like to have a totalizer (like I have now) so I don’t have to crawl into a closet and read out the value for my supplier, also I prefer to have HA read out my daily usage from the Sonoff device and if I’m able to manage that then I’m also able to have Tasmota store it’s usage from the day before. This way I can still read those values even if my HA or my NAS running HA crashes.
![45802CF6-E22B-45F4-AEB9-1CB1BB33D208](https://user-images.githubusercontent.com/42983001/148112496-aa3e98d2-3ced-4fae-9650-6946d148a51f.jpeg)

**Reply by sfromis** [reply]
5V is way outside the ESP8266 specs of 3.3V. May be detrimental to the chip....

I just do not see Sonoff Basic as a good device for using as a sensor platform. Not designed for that, and no good pin availability. Of course, if you *like* to have a messy situation and trouble .... ;-)

**Reply by DutchCoban** [reply]
Until now I don't (yet) smell a chemical burn and it still functions :-)
But you're correct about the specs, I think I'm going to follow the instructions on [this page](https://solarduino.com/how-to-convert-5v-analog-sensor-output-value-to-3-3v-analog-value-for-nodemcu/).

**Reply by sfromis** [reply]
You can slow-fry chips without it getting smelly
